core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/stream/ToMapPerKeyCounter.java

```diff
@@ -0,0 +1,47 @@
+package ai.timefold.solver.core.impl.score.stream;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.BinaryOperator;
+import java.util.stream.Stream;
+
+public final class ToMapPerKeyCounter<Value_> {
+
+    private final Map<Value_, Long> counts = new LinkedHashMap<>(0);
+
+    public long add(Value_ value) {
+        return counts.compute(value, (k, currentCount) -> {
+            if (currentCount == null) {
+                return 1L;
+            } else {
+                return currentCount + 1;
+            }
+        });
+    }
+
+    public long remove(Value_ value) {
+        Long newCount = counts.compute(value, (k, currentCount) -> {
+            if (currentCount > 1L) {
+                return currentCount - 1;
+            } else {
+                return null;
+            }
+        });
+        return newCount == null ? 0L : newCount;
+    }
+
+    public Value_ merge(BinaryOperator<Value_> mergeFunction) {
+        // Rebuilding the value from the collection is not incremental.
+        // The impact is negligible, assuming there are not too many values for the same key.
+        return counts.entrySet()
+                .stream()
+                .flatMap(e -> Stream.generate(e::getKey).limit(e.getValue()))
+                .reduce(mergeFunction)
+                .orElseThrow(() -> new IllegalStateException("Programming error: Should have had at least one value."));
```

Comment:
```suggestion
                .orElseThrow(() -> new IllegalStateException("Impossible state: Should have had at least one value."));
```

"Impossible state" seems to be the convention for this.